File: B05 Inserting/Leetcode57 Insert Interval/InsertInterval.java
/*
Python O( n ) by boundary checking

*/ 

class Solution {
    public int[][] insert(int[][] intervals, int[] newInterval) {
        List<int[]> list = new ArrayList<>();
        int i = 0; 
        
       /* //corner case: 
        if (newInterval[1] < intervals[0][0]){
            
        }
        if (newInterval[0] > intervasl[invervals.length -1][1]){
            
        }
        */
        
        //condition 1: part before overlap
        while(i < intervals.length && intervals[i][1] < newInterval[0]){
            list.add(new int[] {intervals[i][0], intervals[i][1]});
            i++;
        }
        
        
        //condition 2: overlap part
        //int min = 0, max = 0;
        while (i < intervals.length && intervals[i][0] <= newInterval[1] ){
            //min = Math.min(intervals[i][0], newInterval[0]);
            //max = Math.max(intervals[i][1], newInterval[1]);
            newInterval[0] = Math.min(intervals[i][0], newInterval[0]);
            newInterval[1] = Math.max(intervals[i][1], newInterval[1]);
            i++;
        }
        list.add(new int[]{newInterval[0], newInterval[1]});
        //list.add(new int[]{min, max});
        
        //condition 3: parter bybond overlap
        while (i < intervals.length){
            list.add(new int[] {intervals[i][0], intervals[i][1]});
            i++;
        }
        
        int[][]result = new int[list.size()][2];
        for(int j = 0; j <list.size(); j ++){
            result[j][0] = list.get(j)[0];
            result[j][1] = list.get(j)[1];
        }
        return result; 
        
        
    }
    
}

File: B05 Inserting/Leetcode57 Insert Interval/Insert Interval.py
#####Python O( n ) by boundary checking

class Solution:
    def insert(self, intervals, newInterval):
        
        # Constant to help us access start point and end point of interval
        START, END = 0, 1
        
        s, e = newInterval[START], newInterval[END]
        
        left, right = [], []
        
        for cur_interval in intervals:
            
            if cur_interval[END] < s:
                # current interval is on the left-hand side of newInterval
                left += [ cur_interval ]
                
            elif cur_interval[START] > e:
                # current interval is on the right-hand side of newInterval
                right += [ cur_interval ]
                
            else:
                # current interval has overlap with newInterval
                # merge and update boundary
                s = min(s, cur_interval[START])
                e = max(e, cur_interval[END])
                
        return left + [ [s, e] ] + right    

    #####################
    class Solution:
    def insert(self, intervals: List[List[int]], newInterval: List[int]) -> List[List[int]]:
	
        # edge case #1: no intervals in original list
        if len(intervals) == 0:
            return [newInterval]
			
        # edge case #2: new interval goes in the end of the list with no overlap
        if intervals[-1][1] < newInterval[0]:
            return intervals + [newInterval]
			
        # edge case #3: new interval goes in the beginning of the list with no overlap
        if newInterval[1] < intervals[0][0]:
            return [newInterval] + intervals
			
        res = []
        i = 0
        for curr_interval in intervals:
            # new interval doesn't overlap
            if curr_interval[1] < newInterval[0]:
                res.append(curr_interval)
            else:
                first = min(curr_interval[0], newInterval[0])
                last = newInterval[1]
                i = intervals.index(curr_interval)
				
                # while new interval overlaps:
                while i < len(intervals) and last >= intervals[i][0]:
                    last = max(intervals[i][1], newInterval[1])
                    i+=1
                res.append([first, last])
                break
				
        # add the rest
        res += intervals[i:]
        return res